File: src/main/java/cs/ut/jobs/SimulationJob.java
package cs.ut.jobs;

import com.google.common.html.HtmlEscapers;
import cs.ut.config.MasterConfiguration;
import cs.ut.config.items.ModelParameter;
import cs.ut.engine.FileWriter;
import cs.ut.engine.JobManager;
import cs.ut.exceptions.NirdizatiRuntimeException;
import org.apache.commons.io.FilenameUtils;
import org.apache.log4j.Logger;
import org.json.JSONObject;
import org.zkoss.util.resource.Labels;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.nio.file.StandardCopyOption;
import java.util.Map;
import java.util.concurrent.TimeUnit;

public class SimulationJob extends Job {
    private static final Logger log = Logger.getLogger(SimulationJob.class);

    private ModelParameter encoding;
    private ModelParameter bucketing;
    private ModelParameter learner;
    private ModelParameter outcome;

    private File logFile;

    public SimulationJob() {
        super();
    }

    @Override
    public String toString() {
        return FilenameUtils.getBaseName(logFile.getName())
                .concat("_")
                .concat(bucketing.getParameter())
                .concat("_")
                .concat(encoding.getParameter())
                .concat("_")
                .concat(learner.getParameter())
                .concat("_")
                .concat(outcome.getParameter())
                .concat(".pkl");
    }

    @Override
    public void preProcess() {
        JSONObject json = new JSONObject();

        JSONObject params = new JSONObject();
        params.put("max_features", learner.getMaxfeatures() == null ? 0.0 : learner.getMaxfeatures());
        params.put("n_estimators", learner.getEstimators() == null ? 0 : learner.getEstimators());
        params.put("gbm_learning_rate", learner.getGbmrate() == null ? 0 : learner.getGbmrate());
        params.put("n_clusters", 1);

        json.put(outcome.getParameter(),
                new JSONObject().put(bucketing.getParameter().concat("_").concat(encoding.getParameter()),
                        new JSONObject().put(learner.getParameter(), params)));


        FileWriter writer = new FileWriter();
        writer.writeJsonToDisk(json, FilenameUtils.getBaseName(logFile.getName()),
                MasterConfiguration.getInstance().getDirectoryPathConfiguration().getTrainDirectory());
    }

    @Override
    public void execute() {
        try {
            ProcessBuilder pb = new ProcessBuilder("python",
                    "train.py",
                    logFile.getName(),
                    bucketing.getParameter(),
                    encoding.getParameter(),
                    learner.getParameter(),
                    outcome.getParameter());

            pb.directory(new File(coreDir));
            pb.inheritIO();
            Map<String, String> env = pb.environment();
            env.put("PYTHONPATH", scriptDir);

            Process process = pb.start();
            log.debug(pb.command());
            if (!process.waitFor(180, TimeUnit.SECONDS)) {
                process.destroy();
                log.debug("Timed out while trying to create predictor");
            }

            log.debug("Script finished running...");

            File file = new File(scriptDir.concat(pklDir).concat(this.toString()));
            log.debug(file);

            if (!file.exists()) {
                log.debug("Process failed to finish");
            }

            log.debug("Script exited successfully");
        } catch (IOException | InterruptedException e) {
            log.debug("Failed to execute script call", e);
        }
    }

    @Override
    public void postExecute() {
        log.debug(String.format("Trying to move file to user model storage dir <%s>", userModelDir));

        String noExtensionName = FilenameUtils.getBaseName(log.getName());
        File dir = new File(userModelDir.concat(noExtensionName));
        if (!dir.exists() && !dir.mkdir()) {
            log.debug(String.format("Cannot create folder for model with name <%s>", dir.getName()));
        }

        try {
            Files.move(Paths.get(scriptDir.concat(pklDir).concat(this.toString())), Paths.get(userModelDir.concat(noExtensionName.concat("/")).concat(this.toString())), StandardCopyOption.REPLACE_EXISTING);
            log.debug("Successfully moved user model to user model storage directory");

            JobManager.getInstance().getCompletedJobs().add(this);
        } catch (IOException e) {
            log.debug(e);
        }
    }


    public void setEncoding(ModelParameter encoding) {
        this.encoding = encoding;
    }

    public void setBucketing(ModelParameter bucketing) {
        this.bucketing = bucketing;
    }


    public void setLearner(ModelParameter learner) {
        this.learner = learner;
    }

    public ModelParameter getOutcome() {
        return outcome;
    }

    public void setOutcome(ModelParameter outcome) {
        this.outcome = outcome;
    }

    public File getLogFile() {
        return logFile;
    }

    public void setLogFile(File logFile) {
        this.logFile = logFile;
    }

    @Override
    public boolean isNotificationRequired() {
        return true;
    }

    @Override
    public String getNotificationMessage() {
        return Labels.getLabel("job.completed.simulation", new Object[]{logFile.getName()});
    }
}

File: src/main/kotlin/cs/ut/jobs/SimulationJob.kt
package cs.ut.jobs

import cs.ut.configuration.ConfigurationReader
import cs.ut.exceptions.Left
import cs.ut.exceptions.NirdizatiRuntimeException
import cs.ut.exceptions.Right
import cs.ut.exceptions.perform
import cs.ut.jobs.UserRightsJob.Companion.updateACL
import cs.ut.json.JSONHandler
import cs.ut.json.JobInformation
import cs.ut.json.TrainingConfiguration
import cs.ut.providers.Dir
import cs.ut.providers.DirectoryConfiguration
import cs.ut.util.NirdizatiTranslator
import java.io.File


class SimulationJob(
        val configuration: TrainingConfiguration,
        val logFile: File,
        val owner: String,
        id: String = ""
) : Job(id) {

    private var process: Process? = null
    private val configNode = ConfigurationReader.findNode("userPreferences")

    override fun preProcess() {
        log.debug("Generating training parameters for job $this")
        configuration.info = JobInformation(owner, logFile.absolutePath, startTime)

        JSONHandler().writeToFile(configuration, "$id.json", Dir.TRAIN_DIR).apply {
            updateACL(this)
        }
    }

    override fun execute() {
        val python: String = DirectoryConfiguration.dirPath(Dir.PYTHON)
        val parameters = mutableListOf<String>()
        if (configNode.isEnabled()) {
            parameters.add("sudo")
            parameters.add("-u")
            parameters.add(configNode.valueWithIdentifier("userName").value)
        }

        parameters.add(python)
        parameters.add(TRAIN_PY)
        parameters.add(id)


        val execRes = perform {
            val pb = ProcessBuilder(parameters)
            pb.inheritIO()

            pb.directory(File(DirectoryConfiguration.dirPath(Dir.CORE_DIR)))

            val env = pb.environment()
            env["PYTHONPATH"] = DirectoryConfiguration.dirPath(Dir.SCRIPT_DIR)

            log.debug("Script call: ${pb.command()}")
            process = pb.start()

            log.debug("Waiting for process completion")
            process!!.waitFor()
            log.debug("Script finished running...")

            val file = File(DirectoryConfiguration.dirPath(Dir.PKL_DIR)).listFiles().firstOrNull { it.name.contains(this.id) }
            log.debug(file)

            if (file?.exists() == false) {
                status = JobStatus.FAILED
                throw NirdizatiRuntimeException("Script failed to write model to disk, job failed")
            } else {
                log.debug("Script exited successfully")
                process?.destroy()
            }
        }

        when (execRes) {
            is Right -> log.debug("Operation completed successfully")
            is Left -> throw NirdizatiRuntimeException("Script execution failed", execRes.error)
        }
    }

    override fun beforeInterrupt() {
        log.debug("Process ${super.id} has been stopped by the user")
        process?.destroy()
        if (status != JobStatus.COMPLETED) {
            log.debug("Job is not complete -> deleting training file")
            File(DirectoryConfiguration.dirPath(Dir.TRAIN_DIR) + "$id.json").delete()
        }
    }

    override fun onError() {
        super.onError()
        log.debug("Handling error for job $id")
        val file = File(DirectoryConfiguration.dirPath(Dir.TRAIN_DIR) + "$id.json")
        if (file.exists()) {
            log.debug("File ${file.name} exists, deleting file")
            val deleted = file.delete()
            log.debug("File deleted successfully ? $deleted")
        }
        log.debug("Finished handling error for $id")
    }

    override fun isNotificationRequired() = true
    override fun getNotificationMessage() = NirdizatiTranslator.localizeText("job.completed.simulation", this.toString())

    override fun errorOccurred(): Boolean {
        log.debug("Process exit value = ${process?.exitValue()}")
        return process?.exitValue() != 0
    }

    override fun toString(): String {
        return logFile.nameWithoutExtension +
                "_" +
                configuration.bucketing.parameter +
                "_" +
                configuration.encoding.parameter +
                "_" +
                configuration.learner.parameter +
                "_" +
                configuration.outcome.parameter +
                ".pkl"
    }

    private fun convertToNumber(value: String): Number =
            try {
                value.toInt()
            } catch (e: NumberFormatException) {
                value.toDouble()
            }

    companion object {
        const val TRAIN_PY = "train.py"
    }
}